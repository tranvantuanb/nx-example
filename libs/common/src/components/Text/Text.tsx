import React from 'react';
import tw, { styled, css } from 'twin.macro';
import { Typography } from 'antd';
import { TextProps } from 'antd/lib/typography/Text';

const { Text: AntText } = Typography;

export interface CaptionProps {
  size?: 'sm' | 'base' | 'lg' | '2xl';
  uppercase?: boolean;
  children?: React.ReactNode;
  className?: string;
}

const styledTextOptions = {
  shouldForwardProp: (prop: string) => !['uppercase'].includes(prop),
};

const StyledText = styled(AntText, styledTextOptions)<CaptionProps>`
  ${(props) => props.uppercase && tw`uppercase`}

  strong {
    ${tw`font-bold`}
  }
`;

const StyledTextCaption = styled(StyledText)<CaptionProps>`
  ${(props) => props.size === 'sm' && tw`text-sm`}
  ${(props) => props.size === 'base' && tw`text-base`}
  ${(props) => props.size === 'lg' && tw`text-lg`}
  ${(props) => props.size === '2xl' && tw`text-2xl`}
`;

interface IText extends React.FC<TextProps & CaptionProps> {
  Caption: typeof Caption;
}

const Text: IText = ({ children, ...otherProps }) => (
  <StyledText {...otherProps}>{children}</StyledText>
);

const Caption: React.FC<CaptionProps> = ({ children, ...otherProps }) => (
  <StyledTextCaption tw="block" strong {...otherProps}>
    {children}
  </StyledTextCaption>
);

Text.Caption = Caption;

export { Text };
