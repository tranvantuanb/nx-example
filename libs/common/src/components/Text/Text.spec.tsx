import { render } from '../../test';

import { Text } from './Text';

describe('Text', () => {
  it('should render successfully', () => {
    const { container } = render(
      <>
        <Text.Caption size="sm" uppercase>
          Caption 14
        </Text.Caption>
        <Text.Caption size="base" uppercase>
          Caption 16
        </Text.Caption>
        <Text.Caption size="lg" uppercase>
          Caption 18
        </Text.Caption>
        <Text.Caption size="2xl" uppercase>
          Caption 24
        </Text.Caption>
      </>
    );
    expect(container).toBeTruthy();
    expect(container).toMatchSnapshot();
  });
});
