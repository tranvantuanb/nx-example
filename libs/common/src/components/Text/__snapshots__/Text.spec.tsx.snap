// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text should render successfully 1`] = `
.c0 {
  display: block;
}

<div>
  <span
    class="ant-typography c0 css-14xdwh8-StyledText-StyledTextCaption emffahp0"
  >
    <strong>
      Caption 14
    </strong>
  </span>
  <span
    class="ant-typography c0 css-1y6rxd5-StyledText-StyledTextCaption emffahp0"
  >
    <strong>
      Caption 16
    </strong>
  </span>
  <span
    class="ant-typography c0 css-ehnkco-StyledText-StyledTextCaption emffahp0"
  >
    <strong>
      Caption 18
    </strong>
  </span>
  <span
    class="ant-typography c0 css-1f8e10q-StyledText-StyledTextCaption emffahp0"
  >
    <strong>
      Caption 24
    </strong>
  </span>
</div>
`;
